stdlib/public/core/StringUTF8Validation.swift

```diff
@@ -0,0 +1,208 @@
+
+  @inline(__always) func guaranteeIn(_ f: (UInt8) -> Bool) throws {
+    guard let cu = iter.next() else { throw UTF8ValidationError() }
+    guard f(cu) else { throw UTF8ValidationError() }
+  }
+  @inline(__always) func guaranteeContinuation() throws {
+    try guaranteeIn(_isContinuation)
+  }
+
+  func findInvalidRange(_ buf: Slice<UnsafeBufferPointer<UInt8>>) -> Range<Int> {
+    var endIndex = buf.startIndex
+    var iter = buf.makeIterator()
+    _ = iter.next()
+    while let cu = iter.next(), !_isASCII(cu) && !_isUTF8MultiByteLeading(cu) {
+      endIndex += 1
+    }
+    let illegalRange = Range(buf.startIndex...endIndex)
+    assert(illegalRange.clamped(to: (buf.startIndex..<buf.endIndex)) == illegalRange,
+           "illegal range \(illegalRange) out of full range (\(buf.startIndex)..<\(buf.endIndex))")
+    return illegalRange
+  }
+
+  do {
+    var isASCII = true
+    while let cu = iter.next() {
+      if _isASCII(cu) { lastValidIndex += 1; continue }
+      isASCII = false
+      if _slowPath(!_isUTF8MultiByteLeading(cu)) {
+        throw UTF8ValidationError()
+      }
+      // TODO: perhaps a custom branching pattern
+      switch cu {
+      case 0xC2...0xDF:
+        try guaranteeContinuation()
+        lastValidIndex += 2
+      case 0xE0:
+        try guaranteeIn(_isNotOverlong_E0)
+        try guaranteeContinuation()
+        lastValidIndex += 3
+      case 0xE1...0xEC:
+        try guaranteeContinuation()
+        try guaranteeContinuation()
+        lastValidIndex += 3
+      case 0xED:
+        try guaranteeIn(_isNotOverlong_ED)
+        try guaranteeContinuation()
+        lastValidIndex += 3
+      case 0xEE...0xEF:
+        try guaranteeContinuation()
+        try guaranteeContinuation()
+        lastValidIndex += 3
+      case 0xF0:
+        try guaranteeIn(_isNotOverlong_F0)
+        try guaranteeContinuation()
+        try guaranteeContinuation()
+        lastValidIndex += 4
+      case 0xF1...0xF3:
+        try guaranteeContinuation()
+        try guaranteeContinuation()
+        try guaranteeContinuation()
+        lastValidIndex += 4
+      case 0xF4:
+        try guaranteeIn(_isNotOverlong_F4)
+        try guaranteeContinuation()
+        try guaranteeContinuation()
+        lastValidIndex += 4
+      default:
+        fatalError() // TODO: Builtin.unreachable()
+      }
+    }
+    return .success(UTF8ExtraInfo(isASCII: isASCII))
+  } catch {
+    return .error(toBeReplaced: findInvalidRange(buf[lastValidIndex...]))
+  }
+}
+
+@usableFromInline
+func utf8Repair(_ input: UnsafeBufferPointer<UInt8>, firstKnownBrokenRange: Range<Int>) -> String {
+  var x = String()
+  let replacementCharacterCount = utf8ReplacementCharacter.count
+  x.reserveCapacity(input.count + 5 * replacementCharacterCount) // just some extra space
+
+  var lastGoodIndex = 0
+  var brokenRange: Range<Int> = firstKnownBrokenRange
+  var goodRange: Range<Int> = lastGoodIndex ..< brokenRange.startIndex
+  loop: repeat {
```

Comment:
@milseman sorry, this stuff is pretty ugly but should be much faster than the old one as it no longer uses an extra array, so removes an allocation and a copy.

Reply:
@milseman if I clean this up a bit (renaming `x`, and giving the various ranges better names), is this roughly what you thought of?